p2p/net/conn/dial.go

```diff
@@ -92,6 +70,78 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 	return connOut, errOut
 }
 
+// rawConnDial dials the underlying net.Conn + manet.Conns
+func (d *Dialer) rawConnDial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (manet.Conn, error) {
+
+	// before doing anything, check we're going to be able to dial.
+	// we may not support the given address.
+	_, _, err := manet.DialArgs(raddr)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasPrefix(raddr.String(), "/ip4/0.0.0.0") {
+		return nil, debugerror.Errorf("Attempted to connect to zero address: %s", raddr)
+	}
+
+	// get local addr to use.
+	laddr := pickLocalAddr(d.LocalAddrs, raddr)
+
+	log.Debugf("%s dialing %s -- %s --> %s", d.LocalPeer, remote, laddr, raddr)
+	if laddr != nil {
+		// dial using reuseport.Dialer, because we're probably reusing addrs.
+		// this is optimistic, as the reuseDial may fail to bind the port.
+		log.Debugf("trying to reuse: %s", laddr)
+		if nconn, err := d.reuseDial(laddr, raddr); err == nil {
+			// if it worked, wrap the raw net.Conn with our manet.Conn
+			log.Debugf("reuse worked! %s %s %s", laddr, nconn.RemoteAddr(), nconn)
+			return manet.WrapNetConn(nconn)
+		}
+		// if not, we fall back to regular Dial without a local addr specified.
```

Comment:
I originally had a log here, but seemed like it was a lot of noise. I'll remove the "trying to reuse" one.

And, i think it's happening more often than I expect. I think I'm not handling something right and so the `connect()` call fails when it shouldn't have to